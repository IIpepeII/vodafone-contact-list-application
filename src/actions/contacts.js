/*
* Action creators for contacts.
*/

// Handle changes in contactsAreLoading state
export const contactsAreLoading = bool => ({
  type: 'CONTACTS_ARE_LOADING',
  isLoading: bool,
});

// Save initial contacts in contacts state
export const setContacts = contacts => ({
  type: 'SET_CONTACTS',
  contacts,
});

/*
* Fetch contacts asynchronously from the https://randomuser.me API.
*/
export const startSetContacts = () => async (dispatch) => {
  // Set contactsAreLoading state to true.
  dispatch(contactsAreLoading(true));
  const url = 'https://randomuser.me/api/?results=9&nat=us';

  // Every data in the API is lowercase. It is a function to capitalize strings if necessery.
  const capitalize = str => str.charAt(0).toUpperCase() + str.slice(1);

  try {
    // Fetch data
    const response = await fetch(url);
    const data = await response.json();
    const results = await data.results;

    // Create contacts array
    const users = results.map(user => ({
      id: user.id.value,
      firstName: capitalize(user.name.first),
      lastName: capitalize(user.name.last),
      email: user.email,
      cell: user.cell,
    }));

    // Set contactsAreLoading state back to false.
    dispatch(contactsAreLoading(false));

    // Dispatch contacts to Redux store
    dispatch(setContacts(users));
  } catch (e) {
    console.error(e);
  }
};

// Handle delete contact. It is done by id.
export const deleteContact = id => ({
  type: 'DELETE_CONTACT',
  id,
});

// Handle update contact. It is done by id.
export const editContact = (id, updatedData) => ({
  type: 'EDIT_CONTACT',
  id,
  updatedData,
});

// Handle add new contact to the contacts state
export const addContact = contact => ({
  type: 'ADD_CONTACT',
  contact,
});
