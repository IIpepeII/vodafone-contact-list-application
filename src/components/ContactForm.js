/*
* Form for adding and editing contact data.
* Used in ./AddContactPage.js and EditContactPage.js
* props:
* - contact: actual contact data if using the form for editing contact in <EditContactPage />.
* - page: needed to properly render text on form button. 'Save contact' or 'Add contact'.
* - getFormData: function for sending data to parent. <AddContactPage /> or <EditContactPage />
*/
import React, { Component } from 'react';
import {
  Button,
  Form,
  FormGroup,
  Label,
  Input,
} from 'reactstrap';
import PropTypes from 'prop-types';

class ContactForm extends Component {
  static propTypes = {
    contact: PropTypes.objectOf(PropTypes.string),
    page: PropTypes.string.isRequired,
    getFormData: PropTypes.func.isRequired,
  }

  // If form is child of <AddContactPage /> there is no contact prop.
  static defaultProps = {
    contact: {
      firstName: '',
      lastName: '',
      email: '',
      cell: '',
    },
  };

  state = {
    contact: {
      firstName: this.props.contact.firstName,
      lastName: this.props.contact.lastName,
      email: this.props.contact.email,
      cell: this.props.contact.cell,
    },
    error: false,
  }

  // By onchanging form fields update local state by field name property.
  updateContactState = (e) => {
    const { name, value } = e.target;
    this.setState(prevState => ({
      contact: {
        ...prevState.contact,
        [name]: value,
      },
    }));
  }

  // Handle submit form
  handleClickSaveContact = (e) => {
    e.preventDefault();

    const {
      firstName,
      lastName,
      email,
      cell,
    } = this.state.contact;

    /*
    * Regex for email validation.
    */
    const emailReg = /^\w+([\.-]?\w+)*@\w+([\.-]?\w+)*(\.\w{2,3})+$/;

    // Form validation. Check if any field is empty and email address is valid based on regex
    if (firstName.trim().length === 0
      || lastName.trim().length === 0
      || email.trim().length === 0
      || cell.trim().length === 0) {
      this.setState({ error: 'Please, fill out every field!' });
    } else if (!emailReg.test(email)) {
      this.setState({ error: 'Please, enter a valid email address!' });
    } else {
      this.props.getFormData(this.state.contact);
    }
  }

  render() {
    return (
      <Form onSubmit={this.handleClickSaveContact}>
        <FormGroup>
          <Label for="firstName">First name:</Label>
          <Input
            type="text"
            value={this.state.contact.firstName}
            name="firstName"
            placeholder="Enter First Name"
            onChange={this.updateContactState}
          />
        </FormGroup>
        <FormGroup>
          <Label for="lastName">Last name:</Label>
          <Input
            type="text"
            value={this.state.contact.lastName}
            name="lastName"
            placeholder="Enter Last Name"
            onChange={this.updateContactState}
          />
        </FormGroup>
        <FormGroup>
          <Label for="lastName">E-mail:</Label>
          <Input
            type="text"
            value={this.state.contact.email}
            name="email"
            placeholder="Enter Email"
            onChange={this.updateContactState}
          />
        </FormGroup>
        <FormGroup>
          <Label for="cell">Phone:</Label>
          <Input
            type="text"
            value={this.state.contact.cell}
            name="cell"
            placeholder="Enter Phone"
            onChange={this.updateContactState}
          />
        </FormGroup>
        {
          this.state.error && <p className="error">{this.state.error}</p>
        }
        <div className="text-center pt-4">
          <Button className="btn btn--empty btn--gray-empty">
            {this.props.page === 'editContact' ? 'Save Contact' : 'Add Contact'}
          </Button>
        </div>
      </Form>
    );
  }
}

export default ContactForm;
