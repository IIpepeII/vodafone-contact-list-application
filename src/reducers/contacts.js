/*
* Handle actions related to contacts.
*/

/*
* While contacts are fetching from the API, I set this state to true in order to display loader.
* Default state is false.
*/
export const contactsAreLoading = (state = false, action) => {
  switch (action.type) {
    case 'CONTACTS_ARE_LOADING':
      return action.isLoading;
    default:
      return state;
  }
};

/*
* Handle actions related to the contacts state.
* Default state is an empty array.
*/
export const contacts = (state = [], action) => {
  switch (action.type) {
    case 'SET_CONTACTS':
      return action.contacts;
    case 'ADD_CONTACT':
      return [
        action.contact,
        ...state,
      ];
    case 'EDIT_CONTACT':
      return state.map((contact) => {
        if (contact.id === action.id) {
          return {
            ...contact,
            ...action.updatedData,
          };
        }
        return contact;
      });
    case 'DELETE_CONTACT':
      return state.filter(contact => contact.id !== action.id);
    default:
      return state;
  }
};
