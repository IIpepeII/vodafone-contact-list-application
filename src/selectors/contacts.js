/*
* Filter function for testing matches in the last name. Used in components/Dashboard.js
*/
export default (contacts, text) => contacts.filter((contact) => {
  const regex = new RegExp(`^${text}`, 'i');
  return contact.lastName.match(regex);
});
