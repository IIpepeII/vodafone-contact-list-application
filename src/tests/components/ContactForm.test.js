import React from 'react';
import { shallow } from 'enzyme';
import ContactForm from '../../components/ContactForm';
import contacts from '../fixtures/contacts';

let getFormData;

beforeEach(() => {
  getFormData = jest.fn();
});

test('should render correctly on AddContactPage', () => {
  const wrapper = shallow(<ContactForm page="addContact" getFormData={getFormData} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render correctly on EditContactPage', () => {
  const wrapper = shallow(<ContactForm page="editContact" getFormData={getFormData} contact={contacts[0]} />);
  expect(wrapper).toMatchSnapshot();
});
