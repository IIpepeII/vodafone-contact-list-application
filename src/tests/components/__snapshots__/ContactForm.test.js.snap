// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly on AddContactPage 1`] = `
<Form
  onSubmit={[Function]}
  tag="form"
>
  <FormGroup
    tag="div"
  >
    <Label
      for="firstName"
      tag="label"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      First name:
    </Label>
    <Input
      name="firstName"
      onChange={[Function]}
      placeholder="Enter First Name"
      type="text"
      value=""
    />
  </FormGroup>
  <FormGroup
    tag="div"
  >
    <Label
      for="lastName"
      tag="label"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      Last name:
    </Label>
    <Input
      name="lastName"
      onChange={[Function]}
      placeholder="Enter Last Name"
      type="text"
      value=""
    />
  </FormGroup>
  <FormGroup
    tag="div"
  >
    <Label
      for="lastName"
      tag="label"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      E-mail:
    </Label>
    <Input
      name="email"
      onChange={[Function]}
      placeholder="Enter Email"
      type="text"
      value=""
    />
  </FormGroup>
  <FormGroup
    tag="div"
  >
    <Label
      for="cell"
      tag="label"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      Phone:
    </Label>
    <Input
      name="cell"
      onChange={[Function]}
      placeholder="Enter Phone"
      type="text"
      value=""
    />
  </FormGroup>
  <div
    className="text-center pt-4"
  >
    <Button
      className="btn btn--empty btn--gray-empty"
      color="secondary"
      tag="button"
    >
      Add Contact
    </Button>
  </div>
</Form>
`;

exports[`should render correctly on EditContactPage 1`] = `
<Form
  onSubmit={[Function]}
  tag="form"
>
  <FormGroup
    tag="div"
  >
    <Label
      for="firstName"
      tag="label"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      First name:
    </Label>
    <Input
      name="firstName"
      onChange={[Function]}
      placeholder="Enter First Name"
      type="text"
      value="1first"
    />
  </FormGroup>
  <FormGroup
    tag="div"
  >
    <Label
      for="lastName"
      tag="label"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      Last name:
    </Label>
    <Input
      name="lastName"
      onChange={[Function]}
      placeholder="Enter Last Name"
      type="text"
      value="1last"
    />
  </FormGroup>
  <FormGroup
    tag="div"
  >
    <Label
      for="lastName"
      tag="label"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      E-mail:
    </Label>
    <Input
      name="email"
      onChange={[Function]}
      placeholder="Enter Email"
      type="text"
      value="first1.last1@example.com"
    />
  </FormGroup>
  <FormGroup
    tag="div"
  >
    <Label
      for="cell"
      tag="label"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      Phone:
    </Label>
    <Input
      name="cell"
      onChange={[Function]}
      placeholder="Enter Phone"
      type="text"
      value="741852"
    />
  </FormGroup>
  <div
    className="text-center pt-4"
  >
    <Button
      className="btn btn--empty btn--gray-empty"
      color="secondary"
      tag="button"
    >
      Save Contact
    </Button>
  </div>
</Form>
`;
